src/microbit/model/image.py

```diff
@@ -0,0 +1,289 @@
+        sub_arr = []
+
+        for index, elem in enumerate(bytes_translated):
+            if index % width == 0 and index != 0:
+                arr.append(sub_arr)
+                sub_arr = []
+
+            sub_arr.append(elem)
+
+        arr.append(sub_arr)
+        return arr
+
+    # This converts string (with different rows separated by ":")
+    # to 2d array arrangement.
+    def __string_to_array(self, pattern):
+        initial_array, max_subarray_len = self.__string_to_initial_array(pattern)
+
+        # Fill in empty spaces in w x h matrix.
+        for arr_y in initial_array:
+            num_extra_spaces = max_subarray_len - len(arr_y)
+            for _ in range(num_extra_spaces):
+                arr_y.append(0)
+
+        return initial_array
+
+    def __string_to_initial_array(self, pattern):
+        # The result may have spaces in the 2D array
+        # and may unevent sub-array lengths
+        arr = []
+        sub_arr = []
+
+        max_subarray_len = 0
+
+        for elem in pattern:
+            if elem == ":":
+                if len(sub_arr) > max_subarray_len:
+                    max_subarray_len = len(sub_arr)
+                arr.append(sub_arr)
+                sub_arr = []
+            else:
+                sub_arr.append(int(elem))
+
+        if len(pattern) > 0 and not str(pattern)[-1] == ":":
+            arr.append(sub_arr)
+
+        return arr, max_subarray_len
+
+    # This returns the limit if the result is too big.
+    def __limit_result(self, limit, result):
+        if result > limit:
+            return limit
+        else:
+            return result
+
+    def __valid_brightness(self, value):
+        return value >= 0 and value <= 9
+
+    def __valid_pos(self, x, y):
+        return x >= 0 and x < self.width() and y >= 0 and y < self.height()
+
+    def __shift_vertical(self, n):
+
+        res = Image(self.width(), self.height())
+
+        if n > 0:
+            # down
+            res.blit(self, 0, 0, self.width(), self.height() - n, 0, n)
+        else:
+            # up
+            if self.__valid_pos(0, abs(n)):
+                res.blit(self, 0, abs(n), self.width(), self.height() - abs(n), 0, 0)
+
+        return res
+
+    def __shift_horizontal(self, n):
+        res = Image(self.width(), self.height())
+        if n > 0:
+            # right
+            res.blit(self, 0, 0, self.width() - n, self.height(), n, 0)
+        else:
+            # left
+            if self.__valid_pos(abs(n), 0):
+                res.blit(self, abs(n), 0, self.width() - abs(n), self.height(), 0, 0)
+
+        return res
+
+    def __create_string(self):
+        ret_str = ""
+        for index_y in range(0, self.height()):
+            ret_str += self.__row_to_str(index_y)
+        return ret_str
+
+    def __row_to_str(self, y):
+        new_str = ""
+        for x in range(0, self.width()):
+            new_str = new_str + str(self.get_pixel(x, y))
+
+        new_str = new_str + ":"

```

Comment:
nit: keep it consistent with the __create_string function and use new_str += ______________

Reply:
same for in the foreach loop above